README.md
# simple_vcard_parser

A simple and easy to use parser to extract information from a standard vCard string. Major vCard properties from versions 2.1, 3.0, 4.0 are supported.

# Getting started

In the `pubspec.yaml` of your flutter project, add the following
dependency:

```yaml
dependencies:
  ...
  simple_vcard_parser: ^0.1.6
```

In your library add the following import:

```dart
import 'package:simple_vcard_parser/simple_vcard_parser.dart';
```

# Example

```dart
import 'package:simple_vcard_parser/simple_vcard_parser.dart';

String vCardExample40 = '''BEGIN:VCARD
VERSION:4.0
N:Gump;Forrest;;Mr.;
FN:Forrest Gump
ORG:Bubba Gump Shrimp Co.
TITLE:Shrimp Man
PHOTO;MEDIATYPE=image/gif:http://www.example.com/dir_photos/my_photo.gif
TEL;TYPE=work,voice;VALUE=uri:tel:+1-111-555-1212
TEL;TYPE=home,voice;VALUE=uri:tel:+1-404-555-1212
ADR;TYPE=WORK;PREF=1;LABEL="100 Waters Edge\nBaytown\, LA 30314\nUnited States of America":;;100 Waters Edge;Baytown;LA;30314;United States of America
ADR;TYPE=HOME;LABEL="42 Plantation St.\nBaytown\, LA 30314\nUnited States of America":;;42 Plantation St.;Baytown;LA;30314;United States of America
EMAIL;TYPE=INTERNET:forrestgump@example.com
GENDER:M
REV:20080424T195243Z
x-qq:21588891
END:VCARD''';

void main() {
  VCard vc = VCard(vCardExample40);
  print(vc.version); // 4.0
  print(vc.formattedName); // Forrest Gump
  print(vc.organisation); // Bubba Gump Shrimp Co.
  print(vc.title); //Shrimp Man
  print(vc.typedEmail); // [[forrestgump@example.com, [INTERNET]]]
  print(vc
      .typedTelephone); // [[+1-111-555-1212, [VOICE, WORK]], [+1-404-555-1212, [HOME, VOICE]]]
  print(vc.name); //[Gump, Forrest, , Mr.,]
  print(vc.gender); //M
  print(vc
      .typedAddress); // [[[100 Waters Edge, Baytown, LA 30314, United States of America], [WORK]], [[42 Plantation St., Baytown, LA 30314, United States of America], [HOME]]]
  vc.print_lines(); // Will print all vcard lines without start and end tags

  // getWordOfPrefix() can be used to retrieve values from currently unsupported properties
  print(vc.getWordOfPrefix(
      "PHOTO;MEDIATYPE=image/gif:")); //http://www.example.com/dir_photos/my_photo.gif
}

```

# Methods and Properties
### Methods
* getWordOfPrefix (String prefix): returns the value of the provided property prefix without the prefix (first occurence only).
* getWordsOfPrefix (String prefix): similar to getWordOfPrefix, but returns all possible occurences.
* print_lines (): prints formatted vCard lines without start and end tags.

### Properties
* fullString: A vCard representation without START and END tags or any empty lines.
* version: the vCard version.
* name: returns an array containing the components of the name.
* formattedName
* email (Deprecated: refer to typedEmail)
* typedEmail: returns emails along with their types
* organisation
* title
* gender
* typedTelephone: returns an array of telephone numbers along with their type ([[+1-111..., [VOICE, WORK]], [1-404..., [HOME, VOICE]]])
* telephone: returns telephone value if type is not specified in the vCard. (Deprecated: refer to typedTelephone)

## To be supported next:
* Adresses (Done)
* Photos
* Meta-Data

## LICENSE
MIT

CHANGELOG.md
# Changelog

## [0.1.0] - 01-10-2020
* Initial release.

## [0.1.1] - 01-10-2020
* Minor fixes.

## [0.1.2] - 01-10-2020
* email getter fix.

## [0.1.3] - 06-10-2020
* better support for typed telephones (typedTelephone getter).

## [0.1.4] - 12-10-2020
* Added support for gender property.
* Properties such as N and FN (Name and FormattedName) will not crash if it contains any parameters like CHARSET or TYPE.
* telephone getter is deprecated, typedTelephone should be used instead.
* fixed a bug where lowercase tags are not detected. 

## [0.1.5] - 13-10-2020
* Minor bug fix.

## [0.1.6] - 13-10-2020
* Bug fixes.
* Added typedAdress and typedEmail getters.
* Added birthday, note and nickname getters.
* Added print_lines method.
* Deprecated email getter.

## [0.2.0] - 18-08-2022
* Null Safety



example/main.dart
//import 'package:simple_vcard_parser/simple_vcard_parser.dart';

import '../lib/simple_vcard_parser.dart';

String vCardExample40 = '''BEGIN:VCARD
VERSION:4.0
N:Gump;Forrest;;Mr.;
FN:Forrest Gump
ORG:Bubba Gump Shrimp Co.
TITLE:Shrimp Man
PHOTO;MEDIATYPE=image/gif:http://www.example.com/dir_photos/my_photo.gif
TEL;TYPE=work,voice;VALUE=uri:tel:+1-111-555-1212
TEL;TYPE=home,voice;VALUE=uri:tel:+1-404-555-1212
ADR;TYPE=WORK;PREF=1;LABEL="100 Waters Edge\nBaytown\, LA 30314\nUnited States of America":;;100 Waters Edge;Baytown;LA;30314;United States of America
ADR;TYPE=HOME;LABEL="42 Plantation St.\nBaytown\, LA 30314\nUnited States of America":;;42 Plantation St.;Baytown;LA;30314;United States of America
EMAIL;TYPE=INTERNET:forrestgump@example.com
GENDER:M
REV:20080424T195243Z
x-qq:21588891
END:VCARD''';

void main() {
  VCard vc = VCard(vCardExample40);
  print(vc.version); // 4.0
  print(vc.formattedName); // Forrest Gump
  print(vc.organisation); // Bubba Gump Shrimp Co.
  print(vc.title); //Shrimp Man
  print(vc.typedEmail); // [[forrestgump@example.com, [INTERNET]]]
  print(vc
      .typedTelephone); // [[+1-111-555-1212, [VOICE, WORK]], [+1-404-555-1212, [HOME, VOICE]]]
  print(vc.name); //[Gump, Forrest, , Mr.,]
  print(vc.gender); //M
  print(vc
      .typedAddress); // [[[100 Waters Edge, Baytown, LA 30314, United States of America], [WORK]], [[42 Plantation St., Baytown, LA 30314, United States of America], [HOME]]]
  vc.print_lines(); // Will print all vcard lines without start and end tags

  // getWordOfPrefix() can be used to retrieve values from currently unsupported properties
  print(vc.getWordOfPrefix(
      "PHOTO;MEDIATYPE=image/gif:")); //http://www.example.com/dir_photos/my_photo.gif
}


lib/simple_vcard_parser.dart
import 'dart:convert';

class VCard {
  String vCardString;
  late List<String> lines;
  late String version;

  VCard(this.vCardString) {
    lines = LineSplitter().convert(vCardString);
    for (var i = lines.length - 1; i >= 0; i--) {
      if (lines[i].startsWith("BEGIN:VCARD") ||
          lines[i].startsWith("END:VCARD") ||
          lines[i].trim().isEmpty) {
        lines.removeAt(i);
      }
    }

    for (var i = lines.length - 1; i >= 0; i--) {
      if (!lines[i].startsWith(new RegExp(r'^\S+(:|;)'))) {
        String tmpLine = lines[i];
        String prevLine = lines[i - 1];
        lines[i - 1] = prevLine + ', ' + tmpLine;
        lines.removeAt(i);
      }
    }

    version = getWordOfPrefix("VERSION:") ?? "3.0";
  }

  String get fullString => vCardString;

  void print_lines() {
    String s;
    print('lines #${lines.length}');
    for (var i = 0; i < lines.length; i++) {
      s = i.toString().padLeft(2, '0');
      print('$s | ${lines[i]}');
    }
  }

  String? getWordOfPrefix(String prefix) {
    //returns a word of a particular prefix from the tokens minus the prefix [case insensitive]
    for (var i = 0; i < lines.length; i++) {
      if (lines[i].toUpperCase().startsWith(prefix.toUpperCase())) {
        String word = lines[i];
        word = word.substring(prefix.length, word.length);
        return word;
      }
    }
    return null;
  }

  List<String> getWordsOfPrefix(String prefix) {
    //returns a list of words of a particular prefix from the tokens minus the prefix [case insensitive]
    List<String> result = [];

    for (var i = 0; i < lines.length; i++) {
      if (lines[i].toUpperCase().startsWith(prefix.toUpperCase())) {
        String word = lines[i];
        word = word.substring(prefix.length, word.length);
        result.add(word);
      }
    }
    return result;
  }

  String? _strip(String? baseString) {
    if (baseString == null) {
      return null;
    }
    try {
      return RegExp(r'(?<=:).+').firstMatch(baseString)?.group(0);
    } catch (e) {
      return null;
    }
  }

  List<String>? get name => _strip(getWordOfPrefix("N"))?.split(";");
  String? get formattedName => _strip(getWordOfPrefix("FN"));
  String? get nickname => _strip(getWordOfPrefix("NICKNAME"));
  String? get birthday => _strip(getWordOfPrefix("BDAY"));
  String? get organisation => _strip(getWordOfPrefix("ORG"));
  String? get title => _strip(getWordOfPrefix("TITLE"));
  String? get position => _strip(getWordOfPrefix("ROLE"));
  String? get categories => _strip(getWordOfPrefix("CATEGORIES"));
  String? get gender => _strip(getWordOfPrefix('GENDER'));
  String? get note => _strip(getWordOfPrefix('NOTE'));

  List<dynamic> get typedTelephone {
    List<String> telephoneTypes = [
      'TEXT',
      'TEXTPHONE',
      'VOICE',
      'VIDEO',
      'CELL',
      'PAGER',
      'FAX',
      'HOME',
      'WORK',
      'OTHER'
    ];
    List<String> telephones;
    List<String> types = [];
    List<dynamic> result = [];
    String _tel = '';

    telephones = getWordsOfPrefix("TEL");

    for (String tel in telephones) {
      try {
        if (version == "2.1" || version == "3.0") {
          _tel = RegExp(r'(?<=:).+$').firstMatch(tel)?.group(0) ?? '';
        } else if (version == "4.0") {
          _tel = RegExp(r'(?<=tel:).+$').firstMatch(tel)?.group(0) ?? '';
        }
      } catch (e) {
        _tel = '';
      }

      for (String type in telephoneTypes) {
        if (tel.toUpperCase().contains(type)) {
          types.add(type);
        }
      }

      if (_tel.isNotEmpty) {
        result.add([
          _tel,
          types,
        ]);
      }
      _tel = '';
      types = [];
    }

    return result;
  }

  List<dynamic> get typedEmail => _typedProperty('EMAIL');
  List<dynamic> get typedURL => _typedProperty('URL');

  List<dynamic> _typedProperty(String property) {
    // base function for getting typed EMAIL+URL
    List<String> propertyTypes = [
      'HOME',
      'INTERNET',
      'WORK',
      'OTHER',
    ];
    List<String> matches;
    List<String> types = [];
    List<dynamic> result = [];
    String _res = '';

    matches = getWordsOfPrefix(property);

    for (String match in matches) {
      try {
        _res = RegExp(r'(?<=:).+$').firstMatch(match)?.group(0) ?? '';
      } catch (e) {
        _res = '';
      }

      for (String type in propertyTypes) {
        if (match.toUpperCase().contains(type)) {
          types.add(type);
        }
      }

      if (_res.isNotEmpty) {
        if (property == 'ADR') {
          List<String> adress = _res.split(';');
          result.add([
            adress,
            types,
          ]);
        } else {
          result.add([
            _res,
            types,
          ]);
        }
      }
      _res = '';
      types = [];
    }

    return result;
  }

  List<dynamic> get typedAddress {
    List<String> addressTypes = [
      'HOME',
      'WORK',
      'POSTAL',
      'DOM',
    ];
    List<String> addresses;
    List<String> types = [];
    List<dynamic> result = [];
    String _adr = '';

    addresses = getWordsOfPrefix("ADR");

    for (String adr in addresses) {
      try {
        if (version == "2.1" || version == "3.0") {
          _adr = RegExp(r'(?<=(;|:);).+$').firstMatch(adr)?.group(0) ?? '';
        } else if (version == "4.0") {
          _adr = RegExp(r'(?<=LABEL=").+(?=":;)').firstMatch(adr)?.group(0) ?? '';
        }
      } catch (e) {
        _adr = '';
      }

      if (_adr.startsWith(r';')) {
        //remove leading semicolon
        _adr = _adr.substring(1);
      }

      for (String type in addressTypes) {
        if (adr.toUpperCase().contains(type)) {
          types.add(type);
        }
      }

      result.add([
        _adr.split(';'),
        types
      ]); //Add splitted adress ( home;street;city -> [home, street, city]) along with its type
      _adr = '';
      types = [];
    }

    return result; // in this format: [[[adr_1_params], [adr_1_types]], [[adr_2_params], [adr_2_types]]]
  }
}
